frame/evm/precompile/bn128/src/lib.rs

```diff
@@ -92,28 +98,40 @@ impl LinearCostPrecompile for Bn128Mul {
 			sum.y().to_big_endian(&mut buf[32..64]).map_err(|_| ExitError::Other("Cannot fail since 32..64 is 32-byte length".into()))?;
 		}
 
-		Ok((ExitSucceed::Returned, buf.to_vec()))
+		Ok((ExitSucceed::Returned, buf.to_vec(), Bn128Mul::GAS_COST))
 	}
 }
 
 /// The Bn128Pairing builtin
 pub struct Bn128Pairing;
 
-impl LinearCostPrecompile for Bn128Pairing {
-	const BASE: u64 = 15;
-	const WORD: u64 = 3;
+impl Bn128Pairing {
+	// https://eips.ethereum.org/EIPS/eip-1108
+	const BASE_GAS_COST: u64 = 45_000;
+	const GAS_COST_PER_PAIRING: u64 = 34_000;
+}
+
+impl Precompile for Bn128Pairing {
 
 	fn execute(
 		input: &[u8],
-		_: u64,
-	) -> core::result::Result<(ExitSucceed, Vec<u8>), ExitError> {
+		target_gas: Option<u64>,
+		_context: &Context,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, u64), ExitError> {
 		use bn::{AffineG1, AffineG2, Fq, Fq2, pairing_batch, G1, G2, Gt, Group};
 
-		let ret_val = if input.is_empty() {
-			U256::one()
+		let (ret_val, gas_cost) = if input.is_empty() {
+			(U256::one(), Bn128Pairing::BASE_GAS_COST)
 		} else {
 			// (a, b_a, b_b - each 64-byte affine coordinates)
 			let elements = input.len() / 192;
+
+			let gas_cost: u64 =
+				Bn128Pairing::BASE_GAS_COST + (elements as u64 * Bn128Pairing::GAS_COST_PER_PAIRING);
+			if gas_cost > target_gas.expect("Gas limit not provided") { // TODO: Handle
```

Comment:
If target gas is not provided, then it simply means unlimited. In normal contract execution it never happens, so you can just use the good path.

Reply:
Thanks for clarifying :+1: